docs/options/OPTIONS.md
# Peek.js - Options:

Below are all the options that can be passed to the "start()" public function.
<br>
<br>


### Options:

| Type: | Name: | Description: |
| --- | --- | --- |
| *Object* | nodeType | The DOM element types to use the viewer for (can be either an array of strings, or a space-separated string, and defaults to []). |
| *Object* | mode | States the viewing mode to use (1 = CSS, 2 = Attributes, 3 = Size/Position, 4 = Classes. Defaults to 1). |
| *string[]* | showOnly | States the names of the properties that only should be shown (defaults to []). |
| *boolean* | allowEditing | States if editing CSS Properties, Attributes, and Classes is allowed (defaults to false). |
| *boolean* | showIdOrNameInTitle | States if showing the Name/ID of the element in the title of the dialog is enabled (defaults to true). |
| *boolean* | showNodeNameInTitle | States if the node type should be shown in the title (if more than one is being used. Defaults to false). |
| *string[]* | ignoreValues | States the property values that only should be ignored (defaults to []). |
| *boolean* | showLockButtonInTitle | States if the lock symbol button should be shown in the title when the dialog is moved (defaults to true). |
| *number* | offset | States the additional offset to use for the dialog position (defaults to 0). |
| *boolean* | showSearch | States if the search bar should be shown (defaults to true). |
| *number* | showSearchPropertyCount | States minimum number of properties that should be available to show the search bar (defaults to 15). |

<br/>


### Options - Strings:

| Type: | Name: | Description: |
| --- | --- | --- |
| *string* | titleText | An override title to use for the dialog (defaults to null). |

<br/>


## Example:
<br/>

```markdown
<script> 
  $peek.start( {
      nodeType: "button"
  } );
</script>
```

docs/configuration/OPTIONS.md
# Peek.js - Configuration - Options:

Below are all the configuration options that can be passed to the "setConfiguration()" public function.
<br>
<br>


### Options:

| Type: | Name: | Description: |
| --- | --- | --- |
| *number* | dialogShowDelay | States the delay (in milliseconds) that should be used before showing the viewer dialog (defaults to 1000). |
| *number* | dialogHideDelay | States the delay (in milliseconds) that should be used before hiding the viewer dialog when moving away from a DOM element (defaults to 500). |
| *number* | searchDelay | States the delay (in milliseconds) that should be used when typing before searching (defaults to 500). |

<br/>


### Options - Strings:

| Category: | Type: | Name: | Description: |
| --- | --- | --- |
| text | *string* | cssText | The text used for the "CSS" title. |
| text | *string* | attributesText | The text used for the "Attributes" title. |
| text | *string* | sizeText | The text used for the "Size" title. |
| text | *string* | classesText | The text used for the "Classes" title. |
| text | *string* | noAttributesAvailableText | The text used for the "No attributes are available." warning. |
| text | *string* | closeText | The text used for the "Close" button. |
| text | *string* | copySymbolText | The text used for the "❐" copy symbol button. |
| text | *string* | pasteText | The text used for the "Paste" button. |
| text | *string* | pasteSymbolText | The text used for the "☐" paste symbol button. |
| text | *string* | removeText | The text used for the "Remove" button. |
| text | *string* | removeSymbolText | The text used for the "✕" remove symbol button. |
| text | *string* | noClassesAvailableText | The text used for the "No classes are available." warning. |
| text | *string* | searchPropertiesPlaceHolderText | The text used for the "Search properties" placeholder. |
| text | *string* | clearText | The text used for the "Clear" button. |
| text | *string* | clearSymbolText | The text used for the "✕" clear symbol button. |
| text | *string* | noPropertiesFoundForSearchText | The text that should be used for the "No properties were found for your search." search warning. |
| text | *string* | modeNotSupportedText | The text that should be used for the "The mode you have specified is not supported." warning. |
| text | *string* | unknownModeText | The text that should be used for the "Unknown Mode" title. |
| text | *string* | moveUpText | The text that should be used for the "Move Up" button. |
| text | *string* | moveUpSymbolText | The text that should be used for the "Move Up" symbol button. |
| text | *string* | moveDownText | The text that should be used for the "Move Down" button. |
| text | *string* | moveDownSymbolText | The text that should be used for the "Move Down" symbol button. |
| text | *string* | removeElementSymbolText | The text that should be used for the "Remove Element" symbol button. |

<br/>


## Example:
<br/>

```markdown
<script> 
  $peek.setConfiguration( {
      dialogDisplayDelay: 500,
      text: {
        cssText: "CSS Properties"
      }
  } );
</script>
```

docs/CHANGE_LOG.md
# Peek.js - Change Log:

## Version 1.8.1:

#### **Start Options:**
- Added a start option "dialogOffset" (defaults to 0), which states the additional offset to use for the dialogs position.
- Added a start option "showSearch" (defaults to true), which states if the search bar should be shown.
- Added a start option "showSearchPropertyCount" (defaults to 15), which states the minimum number of properties that should be available for the search bar to be visible.

#### **Configuration Options:**
- BREAKING: Renamed the configuration option "dialogDisplayDelay" to "dialogShowDelay".
- BREAKING: Renamed the configuration option "searchDelayDelay" to "searchDelay".
- Added a new configuration option called "dialogHideDelay" (default to 0 milliseconds), which states the delay that should be used before hiding the dialog when moving away from an element.

#### **Adjustments:**
- Moved from ES2016 to ES2020.
- Updated to the latest NPM packages.
- Added an ignore path for ".DS_Store" files when packing the NuGet package.
- Renamed some of the documentation.

#### **Fixes:**
- Fixed some missing documentation in the "options/OPTIONS.md".


## Version 1.7.0:

#### **Start Options:**
- Added a start option "ignoreValues" (defaults to "[]"), which states the property values that only should be ignored.
- Added start option "showLockButtonInTitle" (defaults to true), which states if the lock button should be shown in the title bar.

#### **Configuration Options:**
- Changed the default value for the "text.dialogMovedSymbolText" option to "✸".
- Added start option "lockText" (defaults to "Lock), which states the text to use for the title bar "Lock" button.

#### **UI Improvements:**
- The title bar now uses the main title text as the starting text (instead of the Node Type, when "showNodeNameInTitle" is enabled).
- The Search and Property Value text fields now select all the text when they are focused.
- Added a lock button (configurable) to the right side of the title bar, which will allow you to quickly lock the dialog without having to move it.

#### **Adjustments:**
- Updated to the latest NPM packages.
- Renamed the type "Options" to "StartOptions".
- Added missing type declarations.
- Added configuration settings to force types to be declared.

#### **Fixes:**
- Fixed a fault that prevented a property from being removed when the item was deleted.
- Fixed a fault that caused a double dash to appear in the title bar when "showNodeNameInTitle" is enabled.
- Fixed some script errors in the console when the "stop()" is called.
- Fixed a fault that caused the dialog to still appear for elements after "stop()" is called.

<br>


## Version 1.6.1:
- Fixed missing return types for private shared functions.
- Massive code reorganization (split up into separate folders, renamed, etc).
- NPM package updates.

<br>


## Version 1.6.0:

#### **New Features:**
- Added support to move an element Up/Down within the elements parent node.
- Added support to show when the dialog has been moved and locked.

#### **Configuration:**
- Added configuration option "text.dialogMovedSymbolText" (defaults to "✱"), which states the symbol to show in the dialog title when the dialog has been moved and locked.
- Added configuration option "text.propertyValuePlaceHolderText" (defaults to "Enter value..."), which states the placeholder text to use for the property values.
- Added configuration option "text.modeNotSupportedText" (defaults to "The mode you have specified is not supported."), which states the warning text to show when a mode is not supported.
- Added configuration option "text.unknownModeText" (defaults to "Unknown Mode"), which states the dialog title to use when a mode is not supported.
- Added configuration option "text.moveUpText" (defaults to "Move Up"), which states the text to use for the Move Up button.
- Added configuration option "text.moveUpSymbolText" (defaults to "↑"), which states the text to use for the Move Up symbol button.
- Added configuration option "text.moveDownText" (defaults to "Move Down"), which states the text to use for the Move Down button.
- Added configuration option "text.moveDownSymbolText" (defaults to "↓"), which states the text to use for the Move Down symbol button.
- Added configuration option "text.removeElementSymbolText" (defaults to "⌫"), which states the text to use for the Remove element symbol button.

#### **UI Enhancements:**
- The "Search properties..." search field now uses a new style for the placeholder text.

<br>


## Version 1.5.0:

#### **New Features:**
- Added Search support! This will allow you to filter the properties down via the property name (if more than 15 are available).
- HEX and RGB(a) colors are now shown in the property values are a left border.

#### **Configuration:**
- BREAKING: All text options are now under a new section called "text" (see documentation).
- Added configuration option "text.clearText" (defaults to "Clear"), which states the text to use for the "Clear" button.
- Added configuration option "text.clearSymbolText" (defaults to "✕"), which states the symbol text to use for the "Clear" button.
- Added configuration option "text.searchPropertiesPlaceHolderText" (defaults to "Search properties..."), which states the placeholder text to use for the search input field.
- Added configuration option "text.noPropertiesFoundForSearchText" (defaults to "No properties were found for your search."), which states the warning text to show when no search results are found.
- Added configuration option "searchDelayDelay" (defaults to 500 milliseconds), which states how long the search should wait when typing before running the search.

#### **UI Enhancements:**
- Improved the focus background color used for text input fields.

#### **Adjustments:**
- Moved all string concatenation to use string templates.
- Updated to the latest NPM packages.

#### **Fixes:**

src/ts/options/config.ts
/**
 * Peek.js
 * 
 * A lightweight JavaScript library that attaches a viewer to a specific node type, allowing you to view the CSS properties, attributes, and size/position.
 * 
 * @file        config.ts
 * @version     v1.8.1
 * @author      Bunoon
 * @license     MIT License
 * @copyright   Bunoon 2024
 */


import { type Configuration, type ConfigurationText } from "../type";
import { Default } from "../data/default";


export namespace Config {
    export namespace Options {
        export function get( newConfiguration: Configuration = null! ) : Configuration {
            let configuration: Configuration = Default.getObject( newConfiguration, {} as Configuration );
            configuration.dialogShowDelay = Default.getNumber( configuration.dialogShowDelay, 1000 );
            configuration.dialogHideDelay = Default.getNumber( configuration.dialogHideDelay, 500 );
            configuration.searchDelay = Default.getNumber( configuration.searchDelay, 500 );
    
            configuration = getText( configuration );

            return configuration;
        }
    
        function getText( configuration: Configuration ) : Configuration {
            configuration.text = Default.getObject( configuration.text, {} as ConfigurationText );
            configuration.text!.cssText = Default.getAnyString( configuration.text!.cssText, "CSS" );
            configuration.text!.attributesText = Default.getAnyString( configuration.text!.attributesText, "Attributes" );
            configuration.text!.sizeText = Default.getAnyString( configuration.text!.sizeText, "Size" );
            configuration.text!.classesText = Default.getAnyString( configuration.text!.classesText, "Classes" );
            configuration.text!.noAttributesAvailableText = Default.getAnyString( configuration.text!.noAttributesAvailableText, "No attributes are available." );
            configuration.text!.closeText = Default.getAnyString( configuration.text!.closeText, "Close" );
            configuration.text!.copyText = Default.getAnyString( configuration.text!.copyText, "Copy" );
            configuration.text!.copySymbolText = Default.getAnyString( configuration.text!.copySymbolText, "❐" );
            configuration.text!.pasteText = Default.getAnyString( configuration.text!.pasteText, "Paste" );
            configuration.text!.pasteSymbolText = Default.getAnyString( configuration.text!.pasteSymbolText, "☐" );
            configuration.text!.removeText = Default.getAnyString( configuration.text!.removeText, "Remove" );
            configuration.text!.removeSymbolText = Default.getAnyString( configuration.text!.removeSymbolText, "✕" );
            configuration.text!.noClassesAvailableText = Default.getAnyString( configuration.text!.noClassesAvailableText, "No classes are available." );
            configuration.text!.searchPropertiesPlaceHolderText = Default.getAnyString( configuration.text!.searchPropertiesPlaceHolderText, "Search properties..." );
            configuration.text!.clearText = Default.getAnyString( configuration.text!.clearText, "Clear" );
            configuration.text!.clearSymbolText = Default.getAnyString( configuration.text!.clearSymbolText, "✕" );
            configuration.text!.noPropertiesFoundForSearchText = Default.getAnyString( configuration.text!.noPropertiesFoundForSearchText, "No properties were found for your search." );
            configuration.text!.dialogMovedSymbolText = Default.getAnyString( configuration.text!.dialogMovedSymbolText, "✸" );
            configuration.text!.propertyValuePlaceHolderText = Default.getAnyString( configuration.text!.propertyValuePlaceHolderText, "Enter value..." );
            configuration.text!.modeNotSupportedText = Default.getAnyString( configuration.text!.modeNotSupportedText, "The mode you have specified is not supported." );
            configuration.text!.unknownModeText = Default.getAnyString( configuration.text!.unknownModeText, "Unknown Mode" );
            configuration.text!.moveUpText = Default.getAnyString( configuration.text!.moveUpText, "Move Up" );
            configuration.text!.moveUpSymbolText = Default.getAnyString( configuration.text!.moveUpSymbolText, "↑" );
            configuration.text!.moveDownText = Default.getAnyString( configuration.text!.moveDownText, "Move Down" );
            configuration.text!.moveDownSymbolText = Default.getAnyString( configuration.text!.moveDownSymbolText, "↓" );
            configuration.text!.removeElementSymbolText = Default.getAnyString( configuration.text!.removeElementSymbolText, "⌫" );
            configuration.text!.lockText = Default.getAnyString( configuration.text!.lockText, "Lock" );

            return configuration;
        }
    }
}